galsim/chromatic.py

```diff
@@ -1132,52 +1004,121 @@ def ChromaticAtmosphere(base_obj, base_wavelength, **kwargs):
-                 shift_magnitude*np.cos(parallactic_angle.rad()))
-        return shift
-    ret = ret.shift(shift_fn)
-    return ret
+            raise TypeError("Need to specify zenith_angle and parallactic_angle!")
+
+        # Any remaining kwargs will get forwarded to galsim.dcr.get_refraction
+        # Check that they're valid
+        for kw in kwargs.keys():
+            if kw not in ['temperature', 'pressure', 'H2O_pressure']:
+                raise TypeError("Got unexpected keyword: {0}".format(kw))
+        self.kw = kwargs
+
+        self.base_refraction = galsim.dcr.get_refraction(self.base_wavelength, self.zenith_angle,
+                                                         **kwargs)
+    def __repr__(self):
+        s = 'galsim.ChromaticAtmosphere(%r, base_wavelength=%r, alpha=%r'%(
+                self.base_obj, self.base_wavelength, self.alpha)
+        s += ', zenith_angle=%r, parallactic_angle=%r'%(self.zenith_angle, self.parallactic_angle)
+        for k,v in self.kw.items():
+            s += ', %s=%r'%(k,v)
+        s += ')'
+        return s
+
+    def __str__(self):
+        return 'galsim.ChromaticAtmosphere(%s, base_wavelength=%s, alpha=%s)'%(
+                self.base_obj, self.base_wavelength, self.alpha)
+
+    def build_obj(self):
+        """Build a ChromaticTransformation object for this ChromaticAtmosphere.
+
+        We don't do this right away to help make ChromaticAtmosphere objects be picklable.
+        Building this is quite fast, so we do it on the fly in evaluateAtWavelength and
+        drawImage.
+        """
+        def shift_fn(w):
+            shift_magnitude = galsim.dcr.get_refraction(w, self.zenith_angle, **self.kw)
+            shift_magnitude -= self.base_refraction
+            shift_magnitude = shift_magnitude * (galsim.radians / galsim.arcsec)
+            sinp, cosp = self.parallactic_angle.sincos()
+            shift = (-shift_magnitude * sinp, shift_magnitude * cosp)
+            return shift
+
+        def jac_fn(w):
+            scale = (w/self.base_wavelength)**self.alpha
+            return np.diag([scale, scale])
+
+        frat = lambda w: (w/self.base_wavelength)**(-2.*self.alpha)
+
+        return ChromaticTransformation(self.base_obj, jac=jac_fn, offset=shift_fn, flux_ratio=frat)
+
+    def evaluateAtWavelength(self, wave):
+        """Evaluate this chromatic object at a particular wavelength.
+
+        @param wave     Wavelength in nanometers.
+
+        @returns the monochromatic object at the given wavelength.
+        """
+        return self.build_obj().evaluateAtWavelength(wave)
+
+    def drawImage(self, bandpass, image=None, integrator='trapezoidal', **kwargs):
+        """
+        See ChromaticObject.drawImage for a full description.
+
+        This version usually just calls that one, but if the transformed object has had
+        setupInterpolation called on it, then there is an optimization we can apply here.
```

Comment:
That makes sense to me.  We talk about the optimization in other places that make more sense to me.